bittide/src/Bittide/ClockControl/StabilityChecker.hs

```diff
@@ -0,0 +1,66 @@
+-- SPDX-FileCopyrightText: 2022 Google LLC
+--
+-- SPDX-License-Identifier: Apache-2.0
+{-# LANGUAGE GADTs #-}
+module Bittide.ClockControl.StabilityChecker where
+
+import Clash.Prelude
+
+-- | Strategies to be used with `stabilityChecker`.
+data StabilityStrategy where
+  -- | Remain within a specified @margin@ of a target buffer occupancy for at least
+  -- @cyclesStable@ number of cycles. See 'checkTrackedTargetDeviation' for implementation.
+  TrackedTargetDeviation ::
+    ( 1 <= cyclesStable) =>
+    SNat margin ->
+    SNat cyclesStable ->
+    StabilityStrategy
+
+-- | Checks whether the @Signal@ of buffer occupancies from an elastic buffer is "stable".
+-- Supports different 'StabilityStrategy's.
+stabilityChecker ::
+  forall dom n .
+  (HiddenClockResetEnable dom, KnownNat n) =>
+  -- | Strategy used to determine whether the @Signal@ of buffer occupancies is "stable".
+  StabilityStrategy ->
+  -- | Incoming buffer occupancy.
+  Signal dom (Unsigned n) ->
+  -- | Stability indicator.
+  Signal dom Bool
+stabilityChecker = \case
+  TrackedTargetDeviation maxDev cyclesStable ->
+    checkTrackedTargetDeviation maxDev cyclesStable
+
+-- | Checks if the @Signal@ of buffer occupancies from an elastic buffer is stable.
+-- The @Signal@ is considered stable if it stays within a @margin@ of the target buffer
+-- occupancy for @cyclesStable@ number of cycles. The next target is set to the current
+-- buffer occupancy when the current buffer occupancy is not within margin of
+-- the target.
+checkTrackedTargetDeviation ::
+  forall dom margin cyclesStable n .
+  (HiddenClockResetEnable dom, 1 <= cyclesStable, KnownNat n) =>
+  -- | Maximum number of elements the incoming buffer occupancy is allowed to deviate
+  -- from the current @target@ for it to be considered "stable".
+  SNat margin ->
+  -- | Minimum number of clock cycles the incoming buffer occupancy must remain within the
+  -- @margin@ for it to be considered "stable".
+  SNat cyclesStable ->
+  -- | Incoming buffer occupancy.
+  Signal dom (Unsigned n) ->
+  -- | Stability indicator.
+  Signal dom Bool
+checkTrackedTargetDeviation SNat SNat = mealy go (0, shiftR maxBound 1)
+ where
+  go (cnt, target) input = (newState, isStable)
+   where
+    withinMargin = abs (toSigned target `sub` toSigned input) <= (natToNum @margin)
+
+    newState :: (Index (cyclesStable + 1), Unsigned n)
+    newState
+      | withinMargin = (satSucc SatBound cnt, target)
+      | otherwise    = (0, input)
+
+    isStable = withinMargin && cnt == maxBound
+
+  toSigned :: KnownNat n => Unsigned n -> Signed (n+1)
+  toSigned = bitCoerce . zeroExtend
```

Comment:
`unsignedToSigned` already exists in `Bittide.ClockControl.Callisto.Util`. I think we should move that to `Clash.Sized.Extra` and use that here.

Reply:
Say no more